Spring boot security incorporation

ponomvrenko opened the pull request.

ponomvrenko committed:
spring-boot-starter-security dependency implemented.

ponomvrenko committed:
MapperConfig has become as an interface. SecurityConfig implementation with BCryptPasswordEncoder.

ponomvrenko committed:
Created user entity + implemented with liquibase is_delete column (for soft delete).

security:
  - added CustomUserDetailsService;

ponomvrenko committed:
SecurityConfig with getPasswordEncoder() & securityFilterChain(HttpSecurity http) was configured.

ponomvrenko committed:
Added Role entity.

ponomvrenko committed:
Added Role repository. Also defined public endpoints in SecurityConfig. Added users_roles table. liquibase changes.

BolsunovaNataliia commented on pom.xml:
```diff
@@ -32,7 +32,6 @@
             <groupId>org.springframework.boot</groupId>
```
Good job! But I have a few clarifications.

BolsunovaNataliia commented on src/main/resources/db/changelog/changes/06-create-users-roles-table.yaml:
```diff
@@ -0,0 +1,24 @@
+databaseChangeLog:
```
It's not correct. This table have to consist of 2 columns: user_id, role_id as one privary key

andrii-hoienko replied:
Agree there is no need for separate column for Id, PK of this table could be composed with 2 fields

BolsunovaNataliia commented on src/main/resources/db/changelog/changes/07-insert-data-to-users-roles-table.yaml:
```diff
@@ -0,0 +1,15 @@
+databaseChangeLog:
```
You have to supply query instead hard code ("1" or "2") for user_id and role_id

BolsunovaNataliia commented on src/main/resources/db/changelog/changes/07-insert-data-to-users-roles-table.yaml:
```diff
@@ -0,0 +1,15 @@
+databaseChangeLog:
+  - changeSet:
```
Where is your change set to insert users to users table?

andrii-hoienko commented on src/main/java/com/example/onlinebookstore/model/User.java:
```diff
@@ -24,4 +36,48 @@ public class User {
+    @JoinTable(name = "users_roles",
+            joinColumns = @JoinColumn(name = "user_id"),
+            inverseJoinColumns = @JoinColumn(name = "role_id"))
+    private Set<Role> roles = new HashSet<>();
```
Please exclude this from equals/hashCode and toString

ponomvrenko committed:
Added @ToString.Exclude & @EqualsAndHashCode.Exclude to Set<Role> roles in User model;

ponomvrenko committed:
Modified liquibase changes for improving tables structure:
1. Names for .yaml files were changed;
2. Inserted users with liquibase file;
3. Changed users_roles table structure;
4. Changed file names into db.changelog-master.yaml;

ponomvrenko committed:
Added @EnableMethodSecurity to SecurityConfig

d1sam commented on src/main/resources/db/changelog/changes/08-insert-roles-to-users-roles-table.yaml:
try to improve your code and use valueComputed with preConditions for your insert.
Example:
```
      preConditions:
        - sqlCheck:
            expectedResult: 1
            sql: SELECT COUNT(*) FROM users WHERE email='admin@example.com'
        - sqlCheck:
            expectedResult: 1
            sql: SELECT COUNT(*) FROM roles WHERE name='ROLE_ADMIN'
      changes:
        - insert:
            tableName: user_role
            columns:
              - column:
                  name: user_id
                  valueComputed: (SELECT id FROM users WHERE email='admin@example.com')
              - column:
                  name: role_id
                  valueComputed: (SELECT id FROM roles WHERE name='ROLE_ADMIN')
```

d1sam commented on src/main/java/com/example/onlinebookstore/repository/RoleRepository.java:
```diff
@@ -0,0 +1,7 @@
+import com.example.onlinebookstore.model.Role;
+import org.springframework.data.jpa.repository.JpaRepository;
+
+public interface RoleRepository extends JpaRepository<Role, Long> {
```
```suggestion
@Repository
public interface RoleRepository extends JpaRepository<Role, Long> {
```

d1sam reviewed (changes requested):
Good job, but some changes should be done!

ponomvrenko committed:
Fixes:
- Avoid hardcoding in 8th liquibase script by using sql queries.
- Added @Repository in RoleRepository.
- Removed /api/ from public endpoints cause we already have it as servlet context path.
- Added EntityGraph that helps to avoid LazyInitializationException.

vovasalyha reviewed (approved):
GJ
